File: e2e/graphql/errors_test.go
//go:build e2e

package graphql_test

import (
	"encoding/json"
	"testing"
)

func TestErrors_InvalidSceneID_Query(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`query($id: ID!) {
		scene(id: $id) { id name }
	}`, map[string]any{"id": "nonexistent"})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("scene(nonexistent) returned GraphQL error: %s", gqlResp.Errors[0].Message)
		return
	}

	var result struct {
		Scene *struct {
			ID string `json:"id"`
		} `json:"scene"`
	}
	if err := json.Unmarshal(gqlResp.Data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if result.Scene != nil {
		t.Errorf("expected null scene for nonexistent ID, got %+v", result.Scene)
	}
}

func TestErrors_InvalidAutomationID_Query(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`query($id: ID!) {
		automation(id: $id) { id name }
	}`, map[string]any{"id": "nonexistent"})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("automation(nonexistent) returned GraphQL error: %s", gqlResp.Errors[0].Message)
		return
	}

	var result struct {
		Automation *struct {
			ID string `json:"id"`
		} `json:"automation"`
	}
	if err := json.Unmarshal(gqlResp.Data, &result); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if result.Automation != nil {
		t.Errorf("expected null automation for nonexistent ID, got %+v", result.Automation)
	}
}

func TestErrors_DeleteScene_InvalidID(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`mutation($id: ID!) { deleteScene(id: $id) }`,
		map[string]any{"id": "nonexistent-scene-id"})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("deleteScene(nonexistent) returned error: %s (acceptable)", gqlResp.Errors[0].Message)
		return
	}

	t.Logf("deleteScene(nonexistent) succeeded without error (no-op delete)")
}

func TestErrors_DeleteAutomation_InvalidID(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`mutation($id: ID!) { deleteAutomation(id: $id) }`,
		map[string]any{"id": "nonexistent-automation-id"})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("deleteAutomation(nonexistent) returned error: %s (acceptable)", gqlResp.Errors[0].Message)
		return
	}

	t.Logf("deleteAutomation(nonexistent) succeeded without error (no-op delete)")
}

func TestErrors_DeleteGroup_InvalidID(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`mutation($id: ID!) { deleteGroup(id: $id) }`,
		map[string]any{"id": "nonexistent-group-id"})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("deleteGroup(nonexistent) returned error: %s (acceptable)", gqlResp.Errors[0].Message)
		return
	}

	t.Logf("deleteGroup(nonexistent) succeeded without error (no-op delete)")
}

func TestErrors_AddGroupMember_InvalidType(t *testing.T) {
	data, err := graphqlMutation(`mutation { createGroup(input: { name: "Invalid Type Test" }) { id } }`, nil)
	if err != nil {
		t.Fatalf("create group: %v", err)
	}
	var cr struct {
		CreateGroup struct{ ID string } `json:"createGroup"`
	}
	_ = json.Unmarshal(data, &cr)
	groupID := cr.CreateGroup.ID
	t.Cleanup(func() {
		_, _ = graphqlMutation(`mutation($id: ID!) { deleteGroup(id: $id) }`, map[string]any{"id": groupID})
	})

	mutErr := graphqlMutationExpectError(`mutation($input: AddGroupMemberInput!) {
		addGroupMember(input: $input) { id }
	}`, map[string]any{
		"input": map[string]any{
			"groupId":    groupID,
			"memberType": "invalid",
			"memberId":   "some-id",
		},
	})
	if mutErr != nil {
		t.Fatalf("expected GraphQL error for invalid memberType, got: %v", mutErr)
	}
}

func TestErrors_CreateScene_EmptyName(t *testing.T) {
	gqlResp, err := graphqlPostRaw(`mutation($input: CreateSceneInput!) {
		createScene(input: $input) { id name }
	}`, map[string]any{
		"input": map[string]any{
			"name":    "",
			"actions": []map[string]any{},
		},
	})
	if err != nil {
		t.Fatalf("request failed: %v", err)
	}

	if len(gqlResp.Errors) > 0 {
		t.Logf("createScene with empty name returned error (validated): %s", gqlResp.Errors[0].Message)
	} else {
		var result struct {
			CreateScene struct {
				ID   string `json:"id"`
				Name string `json:"name"`
			} `json:"createScene"`
		}
		if err := json.Unmarshal(gqlResp.Data, &result); err != nil {
			t.Fatalf("unmarshal: %v", err)
		}
		t.Logf("createScene with empty name succeeded — server allows empty names (id=%s)", result.CreateScene.ID)
		t.Cleanup(func() {
			_, _ = graphqlMutation(`mutation($id: ID!) { deleteScene(id: $id) }`, map[string]any{"id": result.CreateScene.ID})
		})
	}
}

File: web/e2e/errors.test.ts
import { describe, it, expect } from "vitest";
import { graphql } from "$lib/gql";
import { getContext } from "./setup.js";

const SCENE_QUERY = graphql(`
  query E2EErrorsScene($id: ID!) {
    scene(id: $id) {
      id
      name
    }
  }
`);

const AUTOMATION_QUERY = graphql(`
  query E2EErrorsAutomation($id: ID!) {
    automation(id: $id) {
      id
      name
    }
  }
`);

const ADD_GROUP_MEMBER = graphql(`
  mutation E2EErrorsAddGroupMember($input: AddGroupMemberInput!) {
    addGroupMember(input: $input) {
      id
    }
  }
`);

const DELETE_SCENE = graphql(`
  mutation E2EErrorsDeleteScene($id: ID!) {
    deleteScene(id: $id)
  }
`);

const CREATE_GROUP = graphql(`
  mutation E2EErrorsCreateGroup($input: CreateGroupInput!) {
    createGroup(input: $input) {
      id
    }
  }
`);

const DELETE_GROUP = graphql(`
  mutation E2EErrorsDeleteGroup($id: ID!) {
    deleteGroup(id: $id)
  }
`);

describe("error handling", () => {
  it("should return error for invalid scene ID", async () => {
    const { graphqlClient } = getContext();

    const result = await graphqlClient.query(SCENE_QUERY, { id: "nonexistent" }).toPromise();

    expect(result.error !== undefined || result.data?.scene === null).toBe(true);
  });

  it("should return error for invalid automation ID", async () => {
    const { graphqlClient } = getContext();

    const result = await graphqlClient.query(AUTOMATION_QUERY, { id: "nonexistent" }).toPromise();

    expect(result.error !== undefined || result.data?.automation === null).toBe(true);
  });

  it("should return error for addGroupMember with invalid memberType", async () => {
    const { graphqlClient } = getContext();

    const group = await graphqlClient
      .mutation(CREATE_GROUP, {
        input: { name: "Error Test Group" },
      })
      .toPromise();
    expect(group.data).toBeDefined();
    const groupId = group.data!.createGroup.id;

    const result = await graphqlClient
      .mutation(ADD_GROUP_MEMBER, {
        input: {
          groupId,
          memberType: "invalid",
          memberId: "some-id",
        },
      })
      .toPromise();

    expect(result.error).toBeDefined();

    await graphqlClient.mutation(DELETE_GROUP, { id: groupId }).toPromise();
  });

  it("should not error for deleteScene with nonexistent ID", async () => {
    const { graphqlClient } = getContext();

    const result = await graphqlClient.mutation(DELETE_SCENE, { id: "nonexistent" }).toPromise();

    expect(result.error).toBeUndefined();
  });
});
